File: 0022-Generate Parentheses/0022-Generate Parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> result;
        string path = "";
        dfs(path, n, n, result);
        return result;
    }
    
private:
    void dfs(string& path, int left, int right, vector<string>& result) {
        if (left == 0 && right == 0) {
            result.push_back(path);
            return;
        }
        
        if (left > 0) {
            path.push_back('(');
            dfs(path, left - 1, right, result);
            path.pop_back();
        }
        
        if (left < right) {
            path.push_back(')');
            dfs(path, left, right - 1, result);
            path.pop_back();
        }
    }
};

File: 0001-0100/0022-Generate Parentheses/0022-Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        def dfs(path, left, right):
            if left == 0 and right == 0:
                result.append(path)
                return
            if left > 0:
                dfs(path + '(', left - 1, right)
            
            if left < right:
                dfs(path + ')', left, right - 1)
        dfs('', n, n)
        return result
